kuma/javascript/src/l10n.js

```diff
@@ -14,71 +14,48 @@
-};
+// This is the type of the string catalog that Django creates for a locale.
+type StringCatalog = { [string]: string | Array<string> };
 
 // The "plural index" function takes a count as its input and returns
 // a number that is an index into an array of plural forms. This default
 // function is suitable for English, but some languages have more than
 // one plural form and require more complicated rules.
-type PluralIndexFunc = number => number;
+type PluralFunction = number => number;
 
 /*
  * These are default values for the state variables that define the
  * current localization.
  */
 const defaultLocale = 'en-US';
-const defaultLocaleData: LocaleData = {
-    catalog: {},
-    plural: null
-};
-const defaultPluralIdx: PluralIndexFunc = n => (+n === 1 ? 0 : 1);
+const defaultStringCatalog: StringCatalog = {};
+const defaultPluralFunction: PluralFunction = n => (+n === 1 ? 0 : 1);
 
 /*
- * The currentLocale, currentLocaleData, and currentPluralIdx variables
- * hold the current state for this module. Their initial values are the
- * defaults above, which are suitable for untranslated English text.
- * The localize() function defines new values for each of them.
+ * The currentLocale, currentStringCatalog, and currentPluralFunction
+ * variables hold the current state for this module. Their initial
+ * values are the defaults above, which are suitable for untranslated
+ * English text.  The localize() function defines new values for each
+ * of them.
  */
 let currentLocale = defaultLocale;
-let currentLocaleData: LocaleData = defaultLocaleData;
-let currentPluralIdx: PluralIndexFunc = defaultPluralIdx;
+let currentStringCatalog: StringCatalog = defaultStringCatalog;
+let currentPluralFunction: PluralFunction = defaultPluralFunction;
 
 /**
  * This function sets up localization for the specified locale.
- * The data argument should be an object in the form returned by
+ * The catalog argument should be a string catalog in the form returned by
  * the JSONCatalog class in
  * https://github.com/django/django/blob/master/django/views/i18n.py
  */
-export function localize(locale: string, data: LocaleData): void {
+export function localize(
+    locale: ?string,
+    catalog: ?StringCatalog,
+    pluralFunction: ?PluralFunction
+): void {
     currentLocale = locale || defaultLocale;
-    currentLocaleData = data || defaultLocaleData;
-
-    // Now define the function that converts numbers to plural forms
-    if (!currentLocaleData.plural) {
-        currentPluralIdx = defaultPluralIdx;
-    } else {
-        // Based on the JS code in
-        // https://github.com/django/django/blob/master/django/views/i18n.py
-        let f = new Function(
-            'n', // Don't change this! The data.plural expression uses it.
-            `let v = ${currentLocaleData.plural};
-             if (v === true) return 1;
-             else if (v === false) return 0;
-             else return v;`
-        );
```

Comment:
This was the problematic code before... The `new Function()` call gets us into CSP trouble.